File: Validate_Binary_Search_Tree.cpp
/*98. Validate Binary Search Tree  QuestionEditorial Solution  My Submissions
Total Accepted: 107794
Total Submissions: 500597
Difficulty: Medium
Given a binary tree, determine if it is a valid binary search tree (BST).

Assume a BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
Example 1:
    2
   / \
  1   3
Binary tree [2,1,3], return true.
Example 2:
    1
   / \
  2   3
Binary tree [1,2,3], return false.*/



/*Set the min_value and max_value
the left child should always less than the root.val
the right child should always larger than the root.val
update the min_value and max_value after each recursive
when check the left child, we set the max value as root.val
when check the right child, we set the min value as root.val*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValid(root, LONG_MIN, LONG_MAX);
    }
    
    bool isValid(TreeNode* root, long min, long max){ // should be "long"
        if(!root) return true;
        if(root->val>=max || root->val<=min) return false;
        // cout << root->val;
        return isValid(root->left, min, root->val) && isValid(root->right, root->val, max);
    }
};





/////////////////////////////////////////////////////////////////////////////////////
//Java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    
    public boolean isValid(TreeNode root, long min, long max){
        if(root==null) return true;
        if(root.val>=max || root.val<=min) return false;
        return isValid(root.left, min, root.val) && isValid(root.right, root.val, max);
    }
}

File: Validate_Binary_Search_Tree.swift
/*
Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.


Example 1:


Input: root = [2,1,3]
Output: true
Example 2:


Input: root = [5,1,4,null,null,3,6]
Output: false
Explanation: The root node's value is 5 but its right child's value is 4.


Constraints:

The number of nodes in the tree is in the range [1, 104].
-231 <= Node.val <= 231 - 1
*/


/*
Solution 2:
DFS iterative

Time Complexity: O(n)
- n is number of nodes in the tree
Space Complexity: O(logn)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        guard let root = root else { return true }
        var stack = [Node]()
        stack.append(Node(root, Int.min, Int.max))

        while !stack.isEmpty {
            let cur = stack.removeLast()
            guard cur.node.val > cur.minVal,
            cur.node.val < cur.maxVal  else {
                return false
            }
            if let l = cur.node.left {
                stack.append(Node(l, cur.minVal, cur.node.val))
            }
            if let r = cur.node.right {
                stack.append(Node(r, cur.node.val, cur.maxVal))
            }
        }
        return true
    }
}

struct Node {
    let node: TreeNode
    let minVal: Int
    let maxVal: Int
    init(_ node: TreeNode, _ minVal: Int, _ maxVal: Int) {
        self.node = node
        self.minVal = minVal
        self.maxVal = maxVal
    }
}

/*
Solution 2:
BFS iterative

Time Complexity: O(n)
- n is number of nodes in the tree
Space Complexity: O(logn)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        guard let root = root else { return true }
        var queue = [Node]()
        queue.append(Node(root, Int.min, Int.max))

        while !queue.isEmpty {
            let cur = queue.removeFirst()
            guard cur.node.val > cur.minVal,
            cur.node.val < cur.maxVal  else {
                return false
            }
            if let l = cur.node.left {
                queue.append(Node(l, cur.minVal, cur.node.val))
            }
            if let r = cur.node.right {
                queue.append(Node(r, cur.node.val, cur.maxVal))
            }
        }
        return true
    }
}

struct Node {
    let node: TreeNode
    let minVal: Int
    let maxVal: Int
    init(_ node: TreeNode, _ minVal: Int, _ maxVal: Int) {
        self.node = node
        self.minVal = minVal
        self.maxVal = maxVal
    }
}

/*
Solution 1: DFS

recursively check if subTree is BST by
setting minVal and maxVal
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        return isValidBST(root, Int.min, Int.max)
    }

    func isValidBST(_ node: TreeNode?, _ minVal: Int, _ maxVal: Int) -> Bool {
        guard let node = node else { return true }
        if node.val >= maxVal || node.val <= minVal { return false }
        return isValidBST(node.left, minVal, node.val)
        && isValidBST(node.right, node.val, maxVal)
    }
}
